### ISSUE TITLE
Show append syntax for SC2207

### ISSUE THREAD

**Issue by danielhoherd** [koalaman/shellcheck#1476]
#### For new checks and feature suggestions
- [X] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
# From within a git repo

files=()

if git config --get remote.origin.url > /dev/null ; then
  if [ "${#@}" -eq 0 ] ; then
    mapfile -t files < <(git ls-files)
  else
    for item in "$@" ; do
      if [ -d "$item" ] ; then
        files+=( $(git ls-files "$item") )
      else
        files+=( "$item" )
      fi
    done
  fi
else
  echo "Not in a git repo, can't continue."
  exit 1
fi
```

#### Here's what shellcheck currently says:

```
In git-show-repo-experts.sh line 40:
        files+=( $(git ls-files "$item") )
                 ^---------------------^ SC2207: Prefer mapfile or read -a to split command output (or quote to avoid splitting).

For more information:
  https://www.shellcheck.net/wiki/SC2207 -- Prefer mapfile or read -a to spli...
```

#### Here's what I wanted or expected to see:

I want the URL <https://www.shellcheck.net/wiki/SC2207> to explain the append syntax of mapfile. `files+=( item )` is an array append, but the syntax described on the linked article `mapfile -t array < <(mycommand)` does not work when appending is required.

**Comment by danielhoherd** [comment]
I found the solution to this problem.

```
files+=( $(git ls-files "$item") )
```
should be changed to
```
mapfile -O "${#files[@]}" -t files <<< "$(git ls-files "$item")"
```

```
-O origin    Begin writing lines to array array at index number origin.
```

**Comment by Mr-Tao** [comment]
@danielhoherd that could result in overwriting of array elements when indices are not contiguous. Also, the syntax is much more opaque. I think that no workaround beats clarity and elegance of `+=` operator and the warning should not be issued for the case of appending array.
I wonder if you can get the maximum index with something less abominable than
```
echo "${!array[@]}" | cut -d" " -f ${#array[@]}
```

**Comment by danielhoherd** [comment]
@Mr-Tao good catch!

```sh
$ for X in {1..5} {10..15} ; do foo[${X}]="foo$X" ; done ;
$ for X in {1..5} {10..15} ; do echo "foo[$X] = ${foo[$X]}" ; done ;
foo[1] = foo1
foo[2] = foo2
foo[3] = foo3
foo[4] = foo4
foo[5] = foo5
foo[10] = foo10
foo[11] = foo11
foo[12] = foo12
foo[13] = foo13
foo[14] = foo14
foo[15] = foo15
$ mapfile -O "${#foo[@]}" -t foo < <(echo {a..z} | xargs -n1 echo)
$ echo "#foo[@] is ${#foo[@]}"
#foo[@] is 32
$ for X in "${!foo[@]}" ; do echo "foo[$X] = ${foo[$X]}" ; done ;
foo[1] = foo1
foo[2] = foo2
foo[3] = foo3
foo[4] = foo4
foo[5] = foo5
foo[10] = foo10
foo[11] = a
foo[12] = b
foo[13] = c
foo[14] = d
foo[15] = e
foo[16] = f
foo[17] = g
foo[18] = h
foo[19] = i
foo[20] = j
foo[21] = k
foo[22] = l
foo[23] = m
foo[24] = n
foo[25] = o
foo[26] = p
foo[27] = q
foo[28] = r
foo[29] = s
foo[30] = t
foo[31] = u
foo[32] = v
foo[33] = w
foo[34] = x
foo[35] = y
foo[36] = z
```

`echo "${!foo[@]}" | awk '{print $NF}'` is only slightly less ugly than `echo "${!array[@]}" | cut -d" " -f ${#array[@]}`

The following works

```sh
$ for X in {1..5} {10..15} ; do foo[${X}]="foo$X" ; done ;
$ mapfile -O "$(echo "${!foo[@]}" | awk '{print $NF + 1}')" -t foo < <(echo {a..z} | xargs -n1 echo)
$ for X in "${!foo[@]}" ; do echo "foo[$X] = ${foo[$X]}" ; done ;
foo[1] = foo1
foo[2] = foo2
foo[3] = foo3
foo[4] = foo4
foo[5] = foo5
foo[10] = foo10
foo[11] = foo11
foo[12] = foo12
foo[13] = foo13
foo[14] = foo14
foo[15] = foo15
foo[16] = a
foo[17] = b
foo[18] = c
foo[19] = d
foo[20] = e
foo[21] = f
foo[22] = g
foo[23] = h
foo[24] = i
foo[25] = j
foo[26] = k
foo[27] = l
foo[28] = m
foo[29] = n
foo[30] = o
foo[31] = p
foo[32] = q
foo[33] = r
foo[34] = s
foo[35] = t
foo[36] = u
foo[37] = v
foo[38] = w
foo[39] = x
foo[40] = y
foo[41] = z
```

Just to append an array... but yet `foo+=( bar1 )` also works, and is inserted as `foo[42]` in this case. Yeah, I think I'll stick with that... 🤦‍♂️
